File: python/remove_all_adjacent_duplicates_in_string.py
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack: list[str] = []
        for i in range(len(s)):
            if len(stack) > 0 and stack[-1] == s[i]:
                stack.pop()
            else:
                stack.append(s[i])
        return "".join(stack)


def test_removeDuplicates_case_1():
    # arrange
    s: str = "abbaca"
    expected: str = "ca"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s)

    # assert
    assert expected == actual


def test_removeDuplicates_case_2():
    # arrange
    s: str = "azxxzy"
    expected: str = "ay"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s)

    # assert
    assert expected == actual

File: rust/src/code/remove_all_adjacent_duplicates_in_string.rs
struct Solution {}

impl Solution {
    pub fn remove_duplicates(s: String) -> String {
        let s: Vec<char> = s.chars().collect();
        let mut stack: Vec<char> = Vec::new();
        for i in 0..s.len() {
            if let Some(&c) = stack.iter().last() {
                if c == s[i] {
                    stack.pop();
                } else {
                    stack.push(s[i]);
                }
                continue;
            }
            stack.push(s[i]);
        }

        stack.iter().collect::<String>()
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn remove_duplicates_case_1() {
        // arrange
        let s: String = "abbaca".to_string();
        let expected: String = "ca".to_string();

        // act
        let actual = Solution::remove_duplicates(s);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn remove_duplicates_case_2() {
        // arrange
        let s: String = "azxxzy".to_string();
        let expected: String = "ay".to_string();

        // act
        let actual = Solution::remove_duplicates(s);

        // assert
        assert_eq!(expected, actual);
    }
}
